### DISCUSSION TITLE
Feature request: Add key bindings to focus on windows by relative direction

### DISCUSSION THREAD

**Discussion by qadzek** [linuxmint/discussions#311] [Ideas]
Cinnamon has a lot of useful key bindings like "Move window to left monitor", "Push tile left" or "Move window to left workspace".

"Focus on window to the left/right/above/below of the active window" would be a nice addition in my opinion.

This was mentioned on the Linux Mint [forums](https://forums.linuxmint.com/viewtopic.php?t=360842) before.
KDE has that feature built in: `System Settings -> Shortcuts -> Global Shortcuts -> KWin -> Switch to Window Above/Below/Left/Right`. Gnome has an [extension](https://extensions.gnome.org/extension/4627/focus-changer/).

**Comment by klangman** [comment]
This sounds like neat feature.. So I created a cinnamon extension last night to implement it.

https://github.com/klangman/Adjacent-Windows

I have done very little testing so far (it's only existed for a few hours after all), but it seems to work well in my environment. 

If you are interested, give it a try and let me know if it works like you expected.

**Reply by qadzek** [reply]
Thanks for creating this extension. Due to your very clear installation instructions, it was easy to install.

I have tested it very briefly. Its basic functionality seems to work perfectly, so I think this would be a nice solution for many users.

However, my workflow often consists of "overlapping" windows. E.g. application A occupies the left part of my screen and application B the right part (thanks to tiling), but beneath A there might be application C. Your extension seems to raise these "hidden" (but not minimized) windows, which is not what I want.

I previously tried the [right-window](https://github.com/ntrrgc/right-window) application, but it has the same issue as your extension does.

I have attempted to solve this myself by writing a Bash script that uses `xdotool mousemove`. It works most of the time, but sometimes fails (probably related to using a multi-monitor setup with different screen scaling). 

Would it be possible to add that feature to your extension, as a setting? A setting like "Only switch to visible windows".

**Reply by klangman** [reply]
I am sure there are a million ways to decide which window is most appropriate to focus, but maybe a simple algorithm is to focus the window in the desired direction that is also highest in the z-order among all the candidate windows? I'll have to think on it a bit after work.

**Reply by klangman** [reply]
I added a new option to use the highest z-order window rather then the closest window. That might work better for your needs?

If you want to try it just update your copy and then restart Cinnamon.

**Reply by qadzek** [reply]
Thanks a lot for this addition; this was indeed what I was looking for.

I will begin using it immediately so I can test it the next couple of days in different configurations (only a laptop, laptop connected to two monitors, monitor scaling enabled/disabled etc.) to give some more feedback.

One bug I noticed when using two monitors:
- when the application layout is like this (a letter signifies an application, a pipe the border between screens and a dot empty space) `|.A|BC|` or like this `|AA|BC|`;
- when window A (on monitor 1) is focused and I want to move to the right (to B, on monitor 2), that B is skipped and it goes straight to C;
- or when C is focused and I want to move to the left, that it jumps immediately to A and skips B.

My settings are: "Allow switching to windows on other monitors" is enabled, "Activate the window that is... Highest in the z-order". 

I temporarily reverted to the previous version of the extension, to test if this bug was already present. Something a bit different happened: when I move to the right from A, at first no window is focused; when I press the key binding again to the right, window C gets focused.

I don't know if you can replicate this. This doesn't seem to happen every time; for example when I manually focus window B it stops doing this.

When using just a single screen, your extension seems to work perfectly already :+1:

EDIT: I just noticed the tooltip "activating the highest window in the z-order might skip over a visible window", so you might already be aware of this issue.

**Reply by klangman** [reply]
Ya.. This is the problem with using only the z-order. If you manually focus 'A' then 'C', then you use the "left" hotkey, the window with the highest z-order in the candidate list of windows will be 'A' since it was the last window to have the focus before 'C' got the focus. Therefore 'A' will get the focus after hitting the left hotkey.

To make better decisions we would need to calculate window viability and use a combination of the visibility of a window along with it's relative position to the currently focused window to select the best candidate window to move the focus to.

I will work on this approach and see if I can get an algorithm that works better.

**Reply by qadzek** [reply]
I had a look at how the [Gnome extension](https://github.com/martinhjartmyr/gnome-shell-extension-focus-changer) implements this feature, but it turns out it's actually an [open issue](https://github.com/martinhjartmyr/gnome-shell-extension-focus-changer/issues/4) in their repository.

Your extension, as it currently stands, seems already helpful for many users who minimize their windows. 

I guess my workflow is to blame here. I open many applications using key bindings, which leads to one window "covering" another. I use [jumpapp](https://github.com/mkropat/jumpapp) to accomplish this. Let's say I press the key binding for Firefox: if a Firefox window is already open, then that window gets the focus; if it isn't, then Firefox is launched.

**Reply by klangman** [reply]
"jumpapp" seems to be a lot like the hotkeys in my CassiaWindowList (https://cinnamon-spices.linuxmint.com/applets/view/372).

I think I can write a better algorithm that is fairly simple but does a reasonable job of taking viability in to account (but it will not be perfect). In reality there will always be some cases where it's not clear what behaviour the user might expect give some partial overlapping windows cases.

**Comment by qadzek** [comment]
The extension was published by @klangman: https://cinnamon-spices.linuxmint.com/extensions/view/99

It has been working great the last couple of weeks on my system.
